locator/src/backup_routes.rs

```diff
@@ -116,22 +123,92 @@ impl BackupRouteProvider for FilesystemRouteProvider {
     }
 }
 
-pub struct GcsRouteProvider {}
+pub struct GcsRouteProvider {
+    bucket_name: String,
+    codec: Codec,
+    object_key: String,
+    client: google_cloud_storage::client::Storage,
+    // If the last cursor does not change, skip upload
+    last_cursor: Mutex<Option<String>>,
+}
 
 impl GcsRouteProvider {
-    pub fn new(_bucket: &str) -> Self {
-        GcsRouteProvider {}
+    pub async fn new(bucket: String) -> Result<Self, BackupError> {
+        let client = google_cloud_storage::client::Storage::builder()
+            .build()
+            .await
+            .map_err(|e| BackupError::GcsInit(e.to_string()))?;
+
+        let bucket_name = format!("projects/_/buckets/{}", &bucket);
+
+        Ok(GcsRouteProvider {
+            bucket_name,
+            codec: Codec::new(Compression::Zstd(1)),
+            object_key: "backup-routes.bin".to_string(),
```

Comment:
Would we have one bucket per locality? Or a shared bucket for all synapse instances?